DPJTH-1 HSM architect changes

jrowden8 opened the pull request:
#### Changes in this pull request
- Take a string for selecting the HSM type.
- Login parameters object.
-  Modify Sample to use the above.
#### Fulfills  [DPJTH-1](https://jira.datalogics.com/browse/DPJTH-1)
#### Checklist for approving this pull request

(PR creator amend this with more conditions if necessary)
- [ ] There are **unit tests** for new/changed code, or a good explanation why this was not possible.
- [ ] All **CI builders** have indicated success (Give them a few minutes to notice the pull request.)
- [ ] The **Pull request title** has the JIRA issue numbers separated by spaces (if any), a space, and then a short, but descriptive summary.
- [ ] **Commit messages** are well formed: [A note about Git commit messages](http://www.tpope.net/node/106)
- [ ] New public packages, classes, and methods are **documented**. (Strongly consider documenting private classes and methods.)
#### Blockers

Add a checkbox for yourself with your **&#64;name** to this list if you are holding this PR open for review. PR Shepherd, please hold off merging this PR until these are all checked:
- [x] _&lt;add your name here with an **@**>_

jrowden8 committed:
HsmManagerFactory: Modify newInstance to accept String for HSM type.

jrowden8 committed:
Remove the run-time IllegalArgumentsException throws clause.

jrowden8 committed:
HsmSignDocument: Fix PMD violation with local of variable declaration.

jrowden8 committed:
HsmLoginParms: Add new class for generic Hsm Login parameters.

jrowden8 committed:
LunaHsmLoginParms: Add new class for Luna Hsm Login parameters.

jrowden8 committed:
HsmManager: Change hsmLogin API to take a HsmLoginParms object.

jrowden8 committed:
LunaHsmManager: Change hsmLogin API to use new HsmLoginParms.

It should use the Luna version of HsmLoginParms (LunaHsmLoginParms).

jrowden8 committed:
LunaHsmManager: Make constants public access.

jrowden8 committed:
HsmManager: Fixup the JavaDoc.

jrowden8 committed:
HsmLoginParms: Add Javadocs.

jrowden8 committed:
LunaHsmLoginParms: Add Javadocs.

jrowden8 committed:
HsmSignDocument: Modify sample to pass String variable to newInstance.

jrowden8 committed:
HsmSignDocument: Modify Sample to pass LunaHsmLoginParms to hsmLogin().

jrowden8 committed:
Move Hsm related source out of the samples package.

jrowden8 committed:
HsmManager: Add new API's to the interface.

jrowden8 committed:
Add Hsm related source to the security package.

jrowden8 committed:
HsmSignDocument: Modify sample to create the credentials in the sample.

jrowden8 committed:
HsmManager: Add new API isLoggedIn.

jrowden8 committed:
LunaHsmManager: Implement isLoggedIn API.

jrowden8 committed:
HsmSignDocument: Utilize isLoggedIn API.

jrowden8 committed:
HsmSignDocument: Make log message severe when cannot login.

jrowden8 committed:
HsmSignDocument: Fix CheckStyle max line 120 chars.

jrowden8 committed:
HsmLoginParms: Remove empty comma.

jrowden8 committed:
HsmManagerFactory: Fixup Javadoc.

jrowden8 committed:
LunaHsmManager: Fixup the Javadoc.

datalogics-robl commented on src/main/java/com/datalogics/pdf/security/LunaHsmManager.java:
```diff
@@ -0,0 +1,180 @@
+    protected LunaHsmManager() throws IOException {
+        super();
+        slotManager = LunaSlotManager.getInstance();
+        lunaKeyStore = getLunaKeyStore();
```
This line doesn't work: you can't obtain the Luna key store until _after_ you are logged in.

datalogics-robl commented on src/main/java/com/datalogics/pdf/security/LunaHsmManager.java:
```diff
@@ -0,0 +1,180 @@
+        try {
+            cert = (X509Certificate) lunaKeyStore.getCertificate(certLabel);
+        } catch (final KeyStoreException e) {
+            throw new SecurityException("Exception while obtaining certificate chain for LunaSa: ", e);
```
- I think all of these getter methods should explicitly check if we are logged in before proceeding, and throw a SecurityException if not. That's likely what happens anyway, but we could return a more detailed message.)
- We should standardize on a spelling for 'Luna SA'

datalogics-robl commented on src/main/java/com/datalogics/pdf/hsm/samples/HsmSignDocument.java:
```diff
@@ -70,30 +77,34 @@ public static void main(final String... args) throws Exception {
+        if (!hsmManager.isLoggedIn()) {
+            // Log in to the HSM
+            hsmManager.hsmLogin(new LunaHsmLoginParms(TOKEN_LABEL, PASSWORD));
+        }
```
I'm doubtful about doing a check if the `hsmManager` is logged in here: since we're in `main()`, we've just procured it, and it's not a singleton, it's guaranteed at this point NOT to be logged in. It certainly doesn't hurt to do it here, but it seems redundant.

Do @jrowden8 or @datalogics-kam have a strong argument for why it SHOULD be checked first?

datalogics-robl commented on src/main/java/com/datalogics/pdf/security/LunaHsmLoginParms.java:
```diff
@@ -0,0 +1,44 @@
+     * @param tokenLbl - The label of the token to which to login
+     * @param pswd - The password to use for the login
+     */
+    public LunaHsmLoginParms(final String tokenLbl, final String pswd) {
```
I feel strongly about not abbreviating the parameter names: let's use `tokenLabel` and `password`.

datalogics-robl commented on src/main/java/com/datalogics/pdf/security/HsmLoginParms.java:
```diff
@@ -0,0 +1,31 @@
+/**
+ * This class represents login parameters that are used to login into any HSM device.
+ */
+public class HsmLoginParms {
```
I'd like this to be fully spelled out: `HsmLoginParameters`.

datalogics-robl commented:
This is all very good work: I like the shape this is taking. Thank you, @jrowden8 !!

jrowden8 committed:
Rename HsmLoginParms to HsmLoginParameters.

jrowden8 committed:
Rename LunaHsmLoginParms to LunaHsmLoginParameters.

jrowden8 committed:
Modify to use renamed HsmLoginParameters & LunaHsmLoginParameters.

jrowden8 committed:
LunaHsmManager: Fixup Exception message.

jrowden8 committed:
Make parameter names explicit.

Commit:
Add cleanup to Javadoc comments

Changes include:
- Removing dash after parameter name
- Prefer 'HSM' and 'Luna SA' spellings

jrowden8 committed:
LunaHsmManager: Add checks that we logged into device beforehand.

jrowden8 committed:
LunaSaManager: Instantiate and load the Keystore after login.

jrowden8 committed:
HsmManagerFactory: Remove throws declaration.

jrowden8 committed:
HsmManager: Modify API's to return more generic instances.

jrowden8 committed:
LunaHsmManager: Change API name to match interface API name.

jrowden8 committed:
HsmSignDocument: Cast to PrivateKey and x509Certificate.

datalogics-robl commented:
Looks great!! 👍

jrowden8 commented:
Thanks @datalogics-robl!
